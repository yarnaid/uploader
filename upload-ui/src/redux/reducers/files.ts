import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { FileType } from "../../types";

const initialState: FileType[] = [];

const filesSlice = createSlice({
  name: "files",
  initialState,
  reducers: {
    loadFilesListRequest: (state) => {
      state;
    },
    setFiles: (state, action: PayloadAction<FileType[]>) => {
      return action.payload;
    },
    resetState: () => initialState,
  },
});

// Export the auto-generated action creators
export const { setFiles, resetState, loadFilesListRequest } =
  filesSlice.actions;

export default filesSlice.reducer;
