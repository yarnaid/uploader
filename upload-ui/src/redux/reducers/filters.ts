import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { FilterType } from "../../types";

const initialState: FilterType[] = [];

const filtersSlice = createSlice({
  name: "filters",
  initialState,
  reducers: {
    setAllFilters: (state, action: PayloadAction<FilterType[]>) => {
      return action.payload;
    },
    removeFilter: (state, action: PayloadAction<FilterType>) => {
      const indexToRemove = state.findIndex(
        (filter) => filter.name === action.payload.name
      );
      if (indexToRemove !== -1) {
        const values = state[indexToRemove].values.filter(
          (value) => !action.payload.values.includes(value)
        );
        if (values.length === 0) {
          state.splice(indexToRemove, 1);
        } else {
          state[indexToRemove].values = values;
        }
      }
    },
    loadAllFilters: (state) => {
      state;
    },
    addFilter: (state, action: PayloadAction<FilterType>) => {
      const index = state.findIndex(
        (filter) => filter.name === action.payload.name
      );
      if (index === -1) {
        state.push(action.payload);
      } else {
        for (const value of action.payload.values) {
          if (!state[index].values.includes(value)) {
            state[index].values.push(value);
          }
        }
      }
    },
  },
});

// Export the auto-generated action creators
export const { setAllFilters, removeFilter, addFilter, loadAllFilters } =
  filtersSlice.actions;

export default filtersSlice.reducer;
