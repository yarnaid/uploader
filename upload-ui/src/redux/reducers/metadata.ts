import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { FileMetadataType, UploadFileType } from "../../types";

const initialState: FileMetadataType = {
  tags: [],
};

const fileSlice = createSlice({
  name: "metadata",
  initialState,
  reducers: {
    setFileMetadata: (state, action: PayloadAction<FileMetadataType>) => {
      return { ...state, ...action.payload };
    },
    updateFileMetadataRequest: (
      state,
      action: PayloadAction<FileMetadataType>
    ) => {
      state = action.payload;
    },
    resetState: () => initialState,
    loadFileInfoRequest: (state, action: PayloadAction<FileMetadataType>) => {
      return { ...state, ...action.payload };
    },
  },
});

// Export the auto-generated action creators
export const {
  setFileMetadata,
  updateFileMetadataRequest,
  resetState,
  loadFileInfoRequest,
} = fileSlice.actions;

export default fileSlice.reducer;
