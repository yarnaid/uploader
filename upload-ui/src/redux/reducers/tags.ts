import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { FilterType } from "../../types";

const initialState: FilterType[] = [];

const fileSlice = createSlice({
  name: "tags",
  initialState,
  reducers: {
    resetState: () => initialState,
    addTag: (state, action: PayloadAction<FilterType>) => {
      const index = state.findIndex((tag) => tag.name === action.payload.name);
      if (index === -1) {
        state.push(action.payload);
      } else {
        state[index].values = [
          ...state[index].values,
          ...action.payload.values,
        ];
      }
    },
    removeTag: (state, action: PayloadAction<FilterType>) => {
      const index = state.findIndex((tag) => tag.name === action.payload.name);
      if (index !== -1) {
        state[index].values = state[index].values.filter(
          (value) => !action.payload.values.includes(value)
        );
      }
    },
    setAllTags: (state, action: PayloadAction<FilterType[]>) => {
      return action.payload;
    },
  },
});

// Export the auto-generated action creators
export const { resetState, addTag, removeTag, setAllTags } = fileSlice.actions;

export default fileSlice.reducer;
