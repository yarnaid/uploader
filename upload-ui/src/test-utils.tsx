import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import { render, RenderOptions } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import configureMockStore from "redux-mock-store";
import { configureStore } from "@reduxjs/toolkit";
import rootReducer from "./redux/reducers";
import {
  FileMetadataType,
  FilterType,
  FileType,
  UploadFileType,
} from "./types";
import type { PreloadedState } from "@reduxjs/toolkit";
import type { RootState, AppStore } from "./redux/store";

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

const mockStore = configureMockStore();

const initialState: {
  files: { files: FileType[] };
  filters: FilterType[];
  file: string;
  tags: FilterType[];
  metadata: FileMetadataType;
} = {
  files: { files: [] },
  filters: [],
  file: "",
  tags: [],
  metadata: { tags: [] },
};

export const store: any = mockStore(initialState);

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <Provider store={store}>
      <MemoryRouter>{children}</MemoryRouter>
    </Provider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({ reducer: rootReducer, preloadedState }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => render(ui, { wrapper: AllTheProviders, ...options });
export * from "@testing-library/react";

export { customRender as render };
