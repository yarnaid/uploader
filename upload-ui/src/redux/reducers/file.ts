import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { setFileMetadata } from "./metadata";

const initialState: string = "";

const fileSlice = createSlice({
  name: "file",
  initialState,
  reducers: {
    setFileData: (state, action: PayloadAction<string>) => {
      return action.payload;
    },
    uploadFileRequest: (state) => {
      return state;
    },
    resetState: () => initialState,
    loadFileInfoRequest: (state, action: PayloadAction<string>) => {
      return state;
    },
  },
});

// Export the auto-generated action creators
export const {
  uploadFileRequest,
  resetState,
  loadFileInfoRequest,
  setFileData,
} = fileSlice.actions;

export default fileSlice.reducer;
