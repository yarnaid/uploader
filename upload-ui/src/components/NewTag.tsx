import React from "react";
import { useState } from "react";
import Button from "react-bootstrap/Button";
import Form from "react-bootstrap/Form";
import Modal from "react-bootstrap/Modal";
import { useDispatch } from "react-redux";
import { addFilter } from "../redux/reducers/filters";

interface NewTagProps {
  name: string;
}

function NewTag({ name }: NewTagProps) {
  const [show, setShow] = useState(false);
  const dispatch = useDispatch();

  const handleClose = () => setShow(false);
  const handleShow = () => setShow(true);
  const [inputValue, setInputValue] = useState("");

  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setInputValue(event.target.value);
  };

  const onAddFilter = () => {
    setShow(false);
    setInputValue("");
    dispatch(addFilter({ name: name, values: [inputValue] }));
  };

  return (
    <>
      <Button variant="success" onClick={handleShow} size="sm">
        +
      </Button>

      <Modal show={show} onHide={handleClose}>
        <Modal.Header closeButton>
          <Modal.Title>Add Tag</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <Form
            onSubmit={() => {
              dispatch(addFilter({ name: name, values: [inputValue] }));
            }}
          >
            <Form.Group className="mb-3" controlId="exampleForm.ControlInput1">
              <Form.Label>
                Enter the value for the tag{" "}
                <code>
                  <b>{name}</b>
                </code>{" "}
                into the input box
              </Form.Label>
              <Form.Control
                type="input"
                placeholder="New tag value..."
                value={inputValue}
                onChange={handleInputChange}
                autoFocus
                // onKeyDown={(event: React.KeyboardEvent<HTMLInputElement>) => {
                //   onAddFilter();
                // }}
              />
            </Form.Group>
          </Form>
        </Modal.Body>
        <Modal.Footer>
          <Button variant="secondary" onClick={handleClose}>
            Close
          </Button>
          <Button variant="primary" onClick={onAddFilter}>
            Save Changes
          </Button>
        </Modal.Footer>
      </Modal>
    </>
  );
}

export default NewTag;
