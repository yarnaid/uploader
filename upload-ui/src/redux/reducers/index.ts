import { combineReducers } from "redux";
import files from "./files";
import file from "./file";
import filters from "./filters";
import tags from "./tags";
import metadata from "./metadata";

const rootReducer = combineReducers({
  files,
  file,
  filters,
  tags,
  metadata,
});

export default rootReducer;
