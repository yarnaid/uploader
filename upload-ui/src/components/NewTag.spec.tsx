import React from "react";
import NewTag from "./NewTag";
import { render, store } from "../test-utils";
import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { addFilter } from "../redux/reducers/filters";

describe("NewTag", () => {
  it("renders", () => {
    const { asFragment } = render(<NewTag name="test" />);
    expect(screen.getByText(/\+/)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
  it("opens modal on click", () => {
    render(<NewTag name="test" />);
    userEvent.click(screen.getByText(/\+/));
    expect(screen.getByText(/Add Tag/i)).toBeInTheDocument();
  });
  it("closes modal on close", () => {
    render(<NewTag name="test" />);
    const addButton = screen.getByText(/\+/i);
    userEvent.click(addButton);
    const closeButton = screen.getByText("Close");
    userEvent.click(closeButton);
    expect(screen.queryByText("Add Filter")).not.toBeInTheDocument();
  });
  it("adds a filter on submit", () => {
    render(<NewTag name="test" />);
    const addButton = screen.getByText(/\+/i);
    userEvent.click(addButton);
    const input = screen.getByPlaceholderText(/New tag value/i);
    userEvent.type(input, "test");
    const submitButton = screen.getByText(/Save Changes/i);
    userEvent.click(submitButton);
    expect(screen.queryByText("Add Filter")).not.toBeInTheDocument();
    const expectedAction = {
      type: addFilter.type,
      payload: { name: "test", values: ["test"] },
    };
    expect(store.getActions()).toEqual([expectedAction]);
  });
});
