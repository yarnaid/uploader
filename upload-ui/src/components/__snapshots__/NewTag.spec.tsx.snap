// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewTag renders 1`] = `
<DocumentFragment>
  <button
    class="btn btn-success btn-sm"
    type="button"
  >
    +
  </button>
</DocumentFragment>
`;
